src/lavinmq/queue/stream_queue_message_store.cr

```diff
@@ -0,0 +1,123 @@
+require "./durable_queue"
+require "../client/channel/consumer"
+
+module LavinMQ
+  class StreamQueue < Queue
+    class StreamQueueMessageStore < MessageStore
+      getter last_offset = 0_i64
+      getter new_messages = Channel(Bool).new
+      @offset_header = ::AMQP::Client::Arguments.new({"x-stream-offset" => 0_i64.as(AMQ::Protocol::Field)})
+
+      # Populate bytesize, size and segment_msg_count
+      private def load_stats_from_segments : Nil
+        last_bytesize = 0_u32
+        @segments.each do |seg, mfile|
+          count = 0u32
+          loop do
+            pos = mfile.pos
+            raise IO::EOFError.new if pos + BytesMessage::MIN_BYTESIZE >= mfile.size # EOF or a message can't fit, truncate
+            ts = IO::ByteFormat::SystemEndian.decode(Int64, mfile.to_slice + pos)
+            break mfile.resize(pos) if ts.zero? # This means that the rest of the file is zero, so resize it
+
+            last_bytesize = bytesize = BytesMessage.skip(mfile)
+            count += 1
+            unless deleted?(seg, pos)
+              @bytesize += bytesize
+              @size += 1
+            end
+          rescue ex : IO::EOFError
+            if mfile.pos < mfile.size
+              Log.warn { "Truncating #{mfile.path} from #{mfile.size} to #{mfile.pos}" }
+              mfile.truncate(mfile.pos)
+            end
+            break
+          rescue ex : OverflowError | AMQ::Protocol::Error::FrameDecode
+            raise Error.new(mfile, cause: ex)
+          end
+
+          begin # sets @last_offset to last message offset
+            mfile.seek(mfile.pos - last_bytesize, IO::Seek::Set)
+            msg = BytesMessage.from_bytes(mfile.to_slice + mfile.pos)
+            @last_offset = offset_from_headers(msg.properties.headers)
+          rescue IndexError
+          end
+
+          mfile.pos = 4
+          @segment_msg_count[seg] = count
+        end
+      end
+
+      def empty?(consumer)
+        return true if @size.zero?
+        @last_offset <= consumer.offset
+      end
+
+      def shift?(consumer) : Envelope? # ameba:disable Metrics/CyclomaticComplexity
+        return if @last_offset <= consumer.offset
+
+        raise ClosedError.new if @closed
+        if sp = consumer.requeued.shift?
+          segment = @segments[sp.segment]
+          begin
+            msg = BytesMessage.from_bytes(segment.to_slice + sp.position)
+            return Envelope.new(sp, msg, redelivered: true)
+          rescue ex
+            raise Error.new(segment, cause: ex)
+          end
+        end
+
+        loop do
+          @rfile_id = consumer.segment
```

Comment:
no point in updating an instance variable here, it will be overwritten by the next consumer
```suggestion
          rfile_id = consumer.segment
```

Reply:
Hmm, yeah I guess. Then we need to overwrite select_next_read_segment from message_store in stream_queue_message_store.